rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,578 @@
+            p,
+            q,
+            d_p,
+            d_q,
+            crt_coefficient,
+            private_exp: private_exponent,
+        })
+    }
+}
+
+impl Drop for RsaKeyPair {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: FFI calls
+            symcrypt_sys::SymCryptRsakeyFree(self.inner());
+        }
+    }
+}
+
+impl RsaPublicKey {
+    /// `set_public_key()` sets only the public key information onto the [`RsaPublicKey`].
+    ///
+    /// `modulus_buffer` takes in a reference to a byte array that contains the modulus of the Rsa key.
+    ///
+    /// `pub_exp` takes in a `&[u8]` that is an array of bytes representing the public exponent.
+    ///
+    /// `num_format` takes in a [`NumberFormat`] that specifies either [`NumberFormat::LSB`] or [`NumberFormat::MSB`] which must match ALL inputs.
+    pub fn set_public_key(
+        n_bits_mod: u32,
+        modulus_buffer: &[u8],
+        pub_exp: &[u8], 
+        num_format: NumberFormat,
+        key_usage: RsaKeyUsage,
+    ) -> Result<Self, SymCryptError> {
+        let rsa_key = allocate_rsa(0, n_bits_mod);
+        let u64_pub_exp = RsaKeyPair::load_msb_first_u64(pub_exp)?;
+        unsafe {
+            // SAFETY: FFI calls
+            // When only setting the public key, ppPrimes, pcbPrimes and nPrimes can be NULL, NULL and 0
+            match symcrypt_sys::SymCryptRsakeySetValue(
+                modulus_buffer.as_ptr(),
+                modulus_buffer.len() as symcrypt_sys::SIZE_T,
+                [u64_pub_exp].as_ptr(),
+                1 as symcrypt_sys::UINT32,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                0,
+                num_format.to_num_format(),
+                key_usage.to_flag(),
+                rsa_key,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(RsaPublicKey {
+                    inner: rsa_key,
+                    key_usage: key_usage,
+                    number_format: num_format,
+                }),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    /// `export_public_key_blob()` will export a [`RsaPublicKey`].
+    pub fn export_public_key_blob(&self) -> Result<RsaPublicKeyBlob, SymCryptError> {
+        let mut modulus_buffer = vec![0u8; self.size_of_modulus() as usize];
+        let mut pub_exp = vec![0u64; 1];
+        unsafe {
+            // SAFETY: FFI calls
+            // When only getting the public key, ppPrimes, pcbPrimes and nPrimes can be NULL, NULL and 0.
+            match symcrypt_sys::SymCryptRsakeyGetValue(
+                self.inner(),
+                modulus_buffer.as_mut_ptr(),
+                modulus_buffer.len() as symcrypt_sys::SIZE_T,
+                pub_exp.as_mut_ptr(),
+                1 as symcrypt_sys::UINT32,
+                ptr::null_mut(),
+                ptr::null_mut(),
+                0,
+                self.number_format.to_num_format(),
```

Comment:
>     self.number_format.to_num_format(), [](http://example.com/codeflow?start=11&length=40)

If LSB needs to be supported, should be input parameter

Reply:
Resolved offline